src/access/examples/ExampleAccessControlDelegate.cpp

```diff
@@ -0,0 +1,1101 @@
+        }
+    }
+
+    void Clear()
+    {
+        mInUse = false;
+        mFabricIndex = chip::kUndefinedFabricIndex;
+        mAuthMode = AuthMode::kPase;
+        mPrivilege = Privilege::kView;
+        memset(mSubjects, SubjectStorage::kEmptyByte, sizeof(mSubjects));
+#if defined(__GNUC__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wclass-memaccess"
+        memset(mTargets, TargetStorage::kEmptyByte, sizeof(mTargets));
+#pragma GCC diagnostic pop
+#else
+        for (auto & target : mTargets)
+        {
+            target.Clear();
+        }
+#endif
+    }
+
+private:
+    static size_t FilterIndex(size_t index, const FabricIndex * fabricIndex)
+    {
+        if (fabricIndex != nullptr)
+        {
+            size_t unfiltered = 0;
+            size_t filtered = 0;
+            for (auto & storage : acl)
+            {
+                if (storage.InUse())
+                {
+                    return ArraySize(acl);
+                }
+                if (storage.mFabricIndex == *fabricIndex)
+                {
+                    if (filtered == index)
+                    {
+                        break;
+                    }
+                    filtered++;
+                }
+                unfiltered++;
+            }
+            return unfiltered;
+        }
+        return index;
+    }
+
+public:
+    static constexpr int kMaxSubjects = 4;
+    static constexpr int kMaxTargets = 3;
+
+private:
+    bool mInUse;
+
+public:
+    FabricIndex mFabricIndex;
+    AuthMode mAuthMode;
+    Privilege mPrivilege;
+    SubjectStorage mSubjects[kMaxSubjects];
+    TargetStorage mTargets[kMaxTargets];
+};
+
+class EntryDelegate : public Entry::Delegate
+{
+public:
+    // Pool support
+    static EntryDelegate pool[kEntryDelegatePoolSize];
+
+    static EntryDelegate * Find(Entry::Delegate & candidate)
+    {
+        if (InPool(candidate))
+        {
+            return &static_cast<EntryDelegate &>(candidate);
+        }
+        for (auto & delegate : pool)
+        {
+            if (!delegate.InUse())
+            {
+                return &delegate;
+            }
+        }
+        return nullptr;
+    }
+
+    static bool InPool(const Entry::Delegate & delegate)
+    {
+        constexpr auto end = pool + ArraySize(pool);
+        return pool <= &delegate && &delegate < end;
+    }
+
+public:
+    void Release() override
+    {
+        assert(mStorage != nullptr);
+        mStorage->Release();
+        mStorage = nullptr;
+    }
+
+    CHIP_ERROR GetAuthMode(AuthMode & authMode) const override
+    {
+        authMode = mStorage->mAuthMode;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR GetFabricIndex(FabricIndex & fabricIndex) const override
+    {
+        fabricIndex = mStorage->mFabricIndex;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR GetPrivilege(Privilege & privilege) const override
+    {
+        privilege = mStorage->mPrivilege;
+        return CHIP_NO_ERROR;
+    }
```

Comment:
If this can never error (or should never error), why do we have a signature that returns error, rather than having easy getter by value return? The EntryDelegate is really an `Entry` logically and therefore if you get an instance of an `Entry` successfully, then all its basic methods should work.

Reply:
It does error in the default case of an uninitialized entry.